src/game/memory.h

```diff
@@ -48,16 +36,90 @@ void *segmented_to_virtual(const void *addr);
 void *virtual_to_segmented(u32 segment, const void *addr);
 void move_segment_table_to_dmem(void);
 
-void main_pool_init(void *start, void *end);
-void *main_pool_alloc(u32 size, u32 side);
-u32 main_pool_free(void *addr);
-void *main_pool_realloc(void *addr, u32 size);
+struct MainPoolRegion {
+    u8* start;
+    u8* end;
+};
+
+#define MAIN_POOL_REGIONS_COUNT 1
+
+struct MainPoolContext {
+    struct MainPoolRegion regions[MAIN_POOL_REGIONS_COUNT];
+};
+
+extern struct MainPoolContext gMainPool;
+
+// takes the first 'size' bytes from 'region'
+static inline void* main_pool_region_try_alloc_from_start(struct MainPoolRegion* region, u32 size) {
+    u8* buf = region->start;
+    u8* newStart = buf + size;
+    if (!buf)
+        __builtin_unreachable();
+
+    region->start = newStart;
+    return buf;
+}
```

Comment:
I recognize this should be well optimized, but I don't like how it can overstep its range without a NULL return or assertion. At minimum, can we iterate over each region at the end of each frame and trip a crash if start ever exceeds end? I worry this may still not be good enough though since at least one frame of gameplay would be processed before the end of the frame, so standard level loading being exceeded would very likely still crash first.

Reply:
A new entrypoint for alloc failures is added `main_pool_alloc_failed`. I am not convinced it will be particularly useful because it is usually too late when alloc fails but it is there.